packages/opentelemetry-plugin-express/src/express.ts

```diff
@@ -0,0 +1,215 @@
+      shimmer.wrap(routerProto, 'use', (original: Function) => {
+        return function use(this: express.Application, arg: string | Function) {
+          const route = original.apply(this, arguments);
+          const layer = this.stack[this.stack.length - 1] as ExpressLayer;
+          plugin._applyPatch(layer, typeof arg === 'string' ? arg : undefined);
+          return route;
+          // tslint:disable-next-line:no-any
+        } as any;
+      });
+      this._logger.debug('patching express.Application.use');
+      shimmer.wrap(
+        this._moduleExports.application,
+        'use',
+        (original: Function) => {
+          return function use(
+            this: { _router: ExpressRouter },
+            arg: string | Function
+          ) {
+            const route = original.apply(this, arguments);
+            const layer = this._router.stack[this._router.stack.length - 1];
+            plugin._applyPatch(
+              layer,
+              typeof arg === 'string' ? arg : undefined
+            );
+            return route;
+            // tslint:disable-next-line:no-any
+          } as any;
+        }
+      );
+    }
+
+    return this._moduleExports;
+  }
+
+  /** Unpatches all MongoDB patched functions. */
+  unpatch(): void {
+    shimmer.unwrap(this._moduleExports.Router.prototype, 'use');
+    shimmer.unwrap(this._moduleExports.Router.prototype, 'route');
+    shimmer.unwrap(this._moduleExports.application, 'use');
+  }
+
+  /**
+   * Store layers path in the request to be able to construct route later
+   * @param request The request where
+   * @param value the value to push into the array
+   */
+  private _storeLayerPath(request: PatchedRequest, value?: string) {
+    if (Array.isArray(request.__ot_middlewares) === false) {
+      Object.defineProperty(request, '__ot_middlewares', {
+        enumerable: false,
+        value: [],
+      });
+    }
+    if (value === undefined) return;
+    (request.__ot_middlewares as string[]).push(value);
+  }
+
+  /** Creates spans for Cursor operations */
+  private _applyPatch(layer: ExpressLayer, layerPath?: string) {
+    const plugin = this;
+    if (layer[kPatched] === true) return;
+    layer[kPatched] = true;
+    this._logger.debug('patching express.Router.Layer.handle');
+    shimmer.wrap(layer, 'handle', function(original: Function) {
+      if (original.length === 4) return original;
+
+      return function(
+        this: ExpressLayer,
+        req: PatchedRequest,
+        res: express.Response,
+        next: express.NextFunction
+      ) {
+        plugin._storeLayerPath(req, layerPath);
+        const route = (req.__ot_middlewares as string[]).join('');
+        const attributes: Attributes = {
+          [AttributeNames.COMPONENT]: plugin._COMPONENT,
+          [AttributeNames.HTTP_ROUTE]: route.length > 0 ? route : undefined,
+        };
+        let spanName = '';
+        if (layer.name === 'router') {
```

Comment:
(Just thinking but) Inside this condition, if the parent span is an http span, we could do something like: ```span.updateName(`${method.toUpperCase()} ${route}`)```. It would be very nice for filtering operations in Zipkin/Jaeger instead of having all routes with values (e.g /user/1, /user/2 etc... but instead /user/:userId), WDYT ?

Reply:
I would prefer to set it as an attribute since both Zipkin/Jeager allows to filter on attributes. Also mutating parent span doesnt feel right